### ISSUE TITLE
Provide PathPatternRequestMatcher.Builder that correctly configures the servlet path

### ISSUE THREAD

**Issue by jzheaux** [spring-projects/spring-boot#37917]
If there is only one dispatcher servlet in the servlet context, then it would be nice to have a simpler way to construct `MvcRequestMatcher` instances.

As it is, an application that is constructing an `MvcRequestMatcher` must do:

```java
@Bean 
SecurityFilterChain appSecurity(HttpSecurity http, HandlerMappingIntrospector introspector) throws Exception {
    MvcRequestMatcher.Builder mvc = new MvcRequestMatcher.Builder(introspector);
    http
        .securityMatchers((security) -> security.requestMatchers(mvc.pattern("/controller/**")))
        /// ...
```

If they have configured `spring.mvc.servlet.path`, then they additionally need to provide that to the matcher as well like so:

```java
@Bean 
SecurityFilterChain appSecurity(HttpSecurity http, HandlerMappingIntrospector introspector) throws Exception {
    MvcRequestMatcher.Builder mvc = new MvcRequestMatcher.Builder(introspector).servletPath("/mvc");
    http
        .securityMatchers((security) -> security.requestMatchers(mvc.pattern("/controller/**")))
        /// ...
```

These two values, the `HandlerMappingIntrospector` and the servlet path configuration, Spring Boot already knows. Given that, it would be nice if a Boot application could instead do:

```java
@Bean 
SecurityFilterChain appSecurity(HttpSecurity http, MvcRequestMatcher.Builder mvc) throws Exception {
    http
        .securityMatchers((security) -> security.requestMatchers(mvc.pattern("/controller/**")))
        /// ...
```

I think it best to publish this bean in the event that there is only one `DispatcherServlet` servlet mapping.

The logic would be something like this (pseudocode follows):

```java
@Conditional(ExactlyOneDispatcherServlet.class)
@Bean 
MvcRequestMatcher.Builder mvcRequestMatcherBuilder(HandlerMappingIntrospector introspector, WebMvcProperties properties) {
    String servletPath = properties.getServlet().getPath();
    MvcRequestMatcher.Builder mvc = new MvcRequestMatcher.Builder(introspector)
    return ("/".equals(servletPath)) ? mvc : mvc.servletPath(servletPath);
}
```

where `ExactlyOneDispatcherServlet` checks the servlet configuration for multiple servlet mappings tied to servlets of type `DispatcherServlet`.

In Spring Security 6.2, the need for an `MvcRequestMatcher` when using `authorizeHttpRequests` [is greatly reduced](https://github.com/spring-projects/spring-security/issues/13568#issuecomment-1759913041). That said, it will still be needed for the foreseeable future when activating the default servlet or when using the `securityMatchers`, `csrf#ignoringRequestMatchers`, and `requiresSecure` DSLs.

**Comment by wilkinsona** [comment]
`MvcRequestMatcher` has been deprecated in favor of `PathPatternRequestMatcher`. @jzheaux does it make sense to do this for `PathPatternRequestMatcher`? Updated pseudo-code:

```java
@Bean
@Conditional(ExactlyOneDispatcherServlet.class)
PathPatternRequestMatcher.Builder patternPatternRequestMatcherBuilder(WebMvcProperties properties) {
    PathPatternRequestMatcher.Builder builder = PathPatternRequestMatcher.withDefaults();
    String servletPath = properties.getServlet().getPath();
    return ("/".equals(servletPath)) ? builder : builder.basePath(servletPath);
}
```

**Comment by jzheaux** [comment]
Yes, that's the idea, @wilkinsona. Though, if there is a `PathPatternParser` in the context, it may be more correct to provide it:

```java
@Bean
@Conditional(ExactlyOneDispatcherServlet.class)
PathPatternRequestMatcher.Builder patternPatternRequestMatcherBuilder(WebMvcProperties properties, PathPatternParser parser) {
    PathPatternRequestMatcher.Builder builder = PathPatternRequestMatcher.withPathPatternParser(parser);
    String servletPath = properties.getServlet().getPath();
    return ("/".equals(servletPath)) ? builder : builder.basePath(servletPath);
}
```

I'll also note for reference that Boot publishes [a `PathPatternRequestMatcher.Builder` bean](https://github.com/spring-projects/spring-boot/blob/4a3cad79831b63902ec8e349ef43484bd28bf29f/module/spring-boot-security/src/main/java/org/springframework/boot/security/autoconfigure/web/servlet/ServletWebSecurityAutoConfiguration.java#L61
) today, the difference being that it publishes it regardless of the number of dispatcher servlets.

/cc @rstoyanchev regarding my use of `PathPatternParser`

**Comment by wilkinsona** [comment]
Thanks, @jzheaux.

I'd forgotten that we're now doing that. It was [added](https://github.com/spring-projects/spring-boot/commit/5e52468e83da9fdb9c519605e9626462e4aeb50a), albeit in a different form, out of necessity when we switched to Security 7.0 snapshots. It's since evolved in https://github.com/spring-projects/spring-boot/commit/05c877d61c97e5c6f62c632cb54aea86f59f5095 and https://github.com/spring-projects/spring-boot/commit/4e425a9b4dedea79142ce4790f0fa4436c000ccf. The latter was in response to https://github.com/spring-projects/spring-security/issues/17746.

I think this leaves us in a good place. The `PathPatternRequestMatcher.Builder` that we define uses the path from the default dispatcher servlet through Boot's `DispatcherServletPath` abstraction. This is essentially a slightly more sophisticated variant of the direct-use-of-`WebMvcProperties` proposal above.
